vfs/directory.go

```diff
@@ -126,3 +161,111 @@ func CreateDirectory(c *Context, doc *DirDoc) (err error) {
 
 	return couchdb.CreateDoc(c.db, doc)
 }
+
+// ModifyDirectoryMetadata modify the metadata associated to a
+// directory. It can be used to rename or move the directory in the
+// VFS.
+func ModifyDirectoryMetadata(c *Context, olddoc *DirDoc, data *DocMetaAttributes) (newdoc *DirDoc, err error) {
+	pth := olddoc.Path
+	name := olddoc.Name
+	tags := olddoc.Tags
+	folderID := olddoc.FolderID
+	mdate := olddoc.UpdatedAt
+
+	if data.FolderID != nil && *data.FolderID != folderID {
+		folderID = *data.FolderID
+		pth, _, err = getFilePath(c, name, folderID)
+		if err != nil {
+			return
+		}
+	}
+
+	if data.Name != "" {
+		name = data.Name
+		pth = path.Join(path.Dir(pth), name)
+	}
+
+	if data.Tags != nil {
+		tags = appendTags(tags, data.Tags)
+	}
+
+	if data.UpdatedAt != nil {
+		mdate = *data.UpdatedAt
+	}
+
+	if mdate.Before(olddoc.CreatedAt) {
+		err = ErrIllegalTime
+		return
+	}
+
+	newdoc, err = NewDirDoc(name, folderID, tags)
+	if err != nil {
+		return
+	}
+
+	newdoc.SetID(olddoc.ID())
+	newdoc.SetRev(olddoc.Rev())
+	newdoc.CreatedAt = olddoc.CreatedAt
+	newdoc.UpdatedAt = mdate
+	newdoc.Path = pth
+
+	if pth != olddoc.Path {
+		err = renameDirectory(olddoc.Path, pth, c.fs)
+		if err != nil {
+			return
+		}
+	}
+
+	err = bulkUpdateDocsPath(c, olddoc, pth)
+	if err != nil {
+		return
+	}
+
+	err = couchdb.UpdateDoc(c.db, newdoc)
+	return
+}
+
+// @TODO remove this method and use couchdb bulk updates instead
+func bulkUpdateDocsPath(c *Context, olddoc *DirDoc, newpath string) error {
+	var children []*DirDoc
+	err := fetchChildren(c, olddoc, FolderDocType, &children)
```

Comment:
I think you are only changing the path of immediate children. It should rename all descendants folder.